runtime/compiler/compile/J9SymbolReferenceTable.cpp

```diff
@@ -897,59 +889,20 @@ J9::SymbolReferenceTable::findOrCreateShadowSymbol(TR::ResolvedMethodSymbol * ow
       containingClass =
          owningMethod->definingClassFromCPFieldRef(comp(), cpIndex, isStatic, &fromResolvedJ9Method);
 
-      if (comp()->compileRelocatableCode())
-         {
-         if (comp()->useTracingBuffer())
-            {
-            if (containingClass == NULL)
-               {
-               const char *bufferPtr = comp()->getTracingBufferCursor();
-               for (char *cursor = comp()->getTracingBufferStart(); cursor < bufferPtr; cursor += TRACING_BUFFER_SEGMENT_SIZE)
-                  {
-                  J9VMThread *vmThread = comp()->fej9()->getCurrentVMThread();
-                  // each subsection *should* be NULL terminated
-                  fprintf(stderr, "%p: %s", vmThread, cursor);
-                  }
-               fprintf(stderr, "End (%d)\n", comp()->getTracingBufferFreeSpace());
-               }
-            comp()->setUseTracingBuffer(false);
-            }
-
-         TR_ASSERT_FATAL(
-            containingClass != NULL,
-            "failed to get defining class of field ref cpIndex=%d in owning method J9Method=%p\n"
-            "\tTR_ResolvedJ9Method::definingClassFromCPFieldRef=%p, TR_ResolvedRelocatableJ9Method::definingClassFromCPFieldRef=%p\n"
-            "\tfromResolvedJ9Method=%p\n"
-            "\tTR_UseSymbolValidationManager=%d",
-            cpIndex,
-            owningMethod->getNonPersistentIdentifier(),
-            owningMethod->TR_ResolvedJ9Method::definingClassFromCPFieldRef(comp(), owningMethod->cp(), cpIndex, isStatic),
-            static_cast<TR_ResolvedRelocatableJ9Method *>(owningMethod)->definingClassFromCPFieldRef(comp(), cpIndex, isStatic),
-            fromResolvedJ9Method,
-            comp()->getOption(TR_UseSymbolValidationManager));
-         }
-      else
-         {
-         TR_ASSERT_FATAL(
-            containingClass != NULL,
-            "failed to get defining class of field ref cpIndex=%d in owning method J9Method=%p",
+      if (containingClass == NULL)
+         comp()->failCompilation<TR::CompilationException>(
```

Comment:
Actually, I wonder if dropping in a comment here that is some variant of your commit message would be useful for documentation purposes:

>findOrCreateShadowSymbol() tries to reuse symrefs produced by findOrFabricateShadowSymbol() when the shadow corresponds to the same field. Detecting that the fields are in fact the same requires checking that the defining class is the same. Normally, the expectation is that if a cp field ref is resolved, then the cp ref pointing to the defining class must also be resolved. However, in #9416, it was discovered that this assumption does not hold in some cases. Under AOT, loading class data from the SCC using findSharedData() may fail, which is propagated to the JIT as a NULL defining class. In addition, if a class is redefined between the point where a field is looked up and where the defining class is looked, the latter may become NULL.
>
>Since the code in question will not work correctly if the defining class is null, this PR tactically replaces the assert with a call to failCompilation().

Reply:
Fixed in force push.

Reply:
Thanks.  LGTM.